### ISSUE TITLE
Ensuring decoration without double decorating in middleware

### ISSUE THREAD

**Issue by trvrnrth** [slimphp/Slim-Http#76]
We have an application that makes use of PSR-15 compliant middleware to handle processing of a ServerRequest and return of a Response.

In each middleware instance process handler I'd like to be able to operate on both the request and response making use of all the helpers added by the Slim decorator.

```
public function process(ServerRequestInterface $request, RequestHandlerInterface $handler): ResponseInterface
{
	// Here I would like to ensure that $request is decorated

	$response = $handler->handle($request);

	// Here I would like to ensure that $response is decorated
}
```

As it currently stands we are in complete control of the creation of the $request and $response instances and can decorate these at source. This may however change in future if we source requests from other providers or make us of third party middleware to generate responses.

Taking the $response decoration scenario the most obvious solution I can think of is to always re-decorate along the lines of:
```
$response = new \Slim\Http\Response($response, $streamFactory);
```

However to my mind this has the following drawbacks:
- Requires extra boilerplate to create/inject a stream factory
- Feels like it goes against the PSR-17 factory pattern
- Will double decorate if $response is already decorated adding extra overhead to every future operation

To avoid the double decoration would require the following, but this further adds to the boilerplate required in each middleware instance:
```
if (!$response instanceof \Slim\Http\Response) {
	$response = new \Slim\Http\Response($response, $streamFactory);
}
```

Do you have any guidance or suggestions for the best way to achieve this with the minimal level of ugliness?

**Comment by lordrhodos** [comment]
Still new to the whole middleware topic, but what about adding this as middlewares itselves? Create a custom `DecorateRequestMiddleware`  and `DecorateResponseMiddleware`  which gets executed after the third party middlewares (beware of the LIFO order middlewares get processed in Slim). You can place them between your *controllable* middleware and the third party middleware thus creating a layer which ensures all middlewares after that get the decorated response and request object.

you my 0.02 cents 😛

**Comment by l0gicgate** [comment]
I would suggest that you create two different middleware for `ServerRequest` and Response respectively. The `ServerRequest` middleware would be placed as the last item of the queue (assuming LIFO order, last in first out) and that the `Response` middleware be placed at the end of the queue (assuming LIFO order again)

This is untested code, it should work in theory although I'm doubtful about the `Response` middleware since you said you're using single pass PSR-15 style middleware.

The `ServerRequest` middleware would look like so:
```php
use Psr\Http\Message\ServerRequestInterface;
use Psr\Http\Server\MiddlewareInterface;
use Psr\Http\Server\RequestHandlerInterface;
use Slim\Http\ServerRequest;

class ServerRequestDecorationDetectionMiddleware implements MiddlewareInterface
{
    public function process(ServerRequestInterface $request, RequestHandlerInterface $handler): ResponseInterface
    {
        if ($request instanceof ServerRequest) {
            $request = new ServerRequest($request);
        }
        return $handler->handle($request);
    }
}
``` 

And the `Response` middleware would look like so:
```php
use Nyholm\Psr7\Factory\Psr17Factory;
use Psr\Http\Message\ServerRequestInterface;
use Psr\Http\Server\MiddlewareInterface;
use Psr\Http\Server\RequestHandlerInterface;
use Slim\Http\Response;

class ResponseDecorationDetectionMiddleware implements MiddlewareInterface
{
    public function process(ServerRequestInterface $request, RequestHandlerInterface $handler): ResponseInterface
    {
        $response = $handler->handle($request);
        if ($response instanceof Response) {
            $psr17Factory = new Psr17Factory();
            $stream = $psr17Factory->createStream();
            $response = new Response($response, $stream);
        }
        return $response;
    }
}
```

**Comment by trvrnrth** [comment]
Thanks for the replies :)

As we ended up wanting a few more helpers I extended the ServerRequest and Response decorators to add those, which led me down the path of needing our own ServerRequestFactory and ResponseFactory which would return instances of that type.

Given I then had the custom factories anyway I added in a decorate function to each:
```
public function decorate(ResponseInterface $response): ResponseInterface
{
	if ($response instanceof Response) {
		return $response;
	}
	return new Response($response, $this->streamFactory);
}
```
```
public function decorate(ServerRequestInterface $request): ServerRequestInterface
{
	if ($request instanceof ServerRequest) {
		return $request;
	}
	return new ServerRequest($request);
}
```
I can then call this anywhere I specifically need a decorated version of the request or response when I am not otherwise sure it is already decorated. So, taking the middleware I can decorate the response as needed (the response factory is pulled in from our DI container in the constructor):
```
public function process(ServerRequestInterface $request, RequestHandlerInterface $handler): ResponseInterface
{
	// Actions

	$response = $this->responseFactory->decorate(
		$handler->handle($request)
	);

	// Actions

	return $response;
}
```
This does still add to the boilerplate in each middleware but as it's now a function call to the factory rather than a copy-paste job I'm happier with it.

For the server request side of things as I know the request is already decorated in our app, for any middleware that we're not concerned about being portable I can save the need for the call and just type-hint accordingly to keep the analysers and IDEs happy:
```
/**
 * {@inheritdoc}
 *
 * @param ServerRequestContract $request
 */
public function process($request, RequestHandlerInterface $handler): ResponseInterface
{
	// Do the things
}
```

As it stands we're happy enough with this solution and it's nice to see that it's generally in-line with your thoughts and recommendations as well. Thanks again for taking the time to reply.
